CounterExamples

Dargones opened the pull request:
# Preview

Counterexample feature can currently only extract values of `int` and `bool` types. This PR adds/resurrects support for:

1) Reals, chars, and bit-vectors:
<img width="832" alt="Basics" src="https://user-images.githubusercontent.com/21084554/127909929-ed773f1c-ae09-4cd6-9c06-8ea86f925ce7.png">

2) Sets and sequences, including strings:
<img width="1019" alt="Seqs Sets" src="https://user-images.githubusercontent.com/21084554/127909911-4df37883-a0e7-40cb-97c4-2753296e0cf3.png">

3) Datatypes:
<img width="1034" alt="Datatypes" src="https://user-images.githubusercontent.com/21084554/127973963-1b771e16-f03c-48c4-b4f1-92fc5980df4c.png">

4) Objects of user-defined types. Fields can be extracted recursively up to some desired limit:
<img width="801" alt="Objects" src="https://user-images.githubusercontent.com/21084554/127973938-48f1e14d-639d-4aed-b1c9-600956ad557b.png">

5) Arrays and multi-dimensional arrays: 
<img width="837" alt="Arrays" src="https://user-images.githubusercontent.com/21084554/127904122-aef29163-38ec-4b37-88a3-85bb77b52e6d.png">

6) Type extraction. You can now see the type of every variable in the counterexample, which can be useful for downstream tasks (such as automatic test generation):
<img width="1167" alt="TypeExtraction" src="https://user-images.githubusercontent.com/21084554/127913090-2aac7c31-8fa5-487c-9da0-32dc3cc6340e.png">

# The Changes

The changes I propose mostly concern the files in `DafnyServer/boogie` directory (renamed to `DafnyServer/CounterExampleGeneration`). I think it makes sense to group all these changes together because they are tightly interrelated but I also realize that this is a lot to review, so let me know if I should break this up into multiple PRs. Every commit should be stable and passing all tests. Below is a more detailed breakdown of the changes:

1) Merge together [LanguageModel](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/Namer.cs#L19), [DafnyModel](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L28), and [Provider](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L11). Call the resulting class `DafnyModel` and put it into a file with a corresponding name. 
2) Merge together [DisplayNode](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DataModel.cs#L39), [FieldNode](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L392), [ElementNode](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L370), and [VariableNode](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L400). Call the new class `DafnyModelVariable` and put it into a file with a corresponding name. 
3) Separate the [DafnyModelState](https://github.com/dafny-lang/dafny/blob/ed2f869a2025f95dd3176b5825faeeffbe6efa1e/Source/DafnyServer/boogie/DafnyProvider.cs#L304) class into its own file and transfer to it some of the functionality previously included in the `DafnyModel` (such as the function responsible for encoding the name of the model state).
4) Add support for types and collections described above. Doing so requires extensive fixes and additions to `CanonicalName` and `GetExpansion` methods in `DafnyModel`. Adding the ability to extract types requires a slew of new methods also defined in `DafnyModel` along with the changes to how the `Value` property of `DafnyModelVariable` is computed.
5) Make smaller changes to [CounterExampleProvider](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyServer/CounterExampleProvider.cs) (incidentally, should this earlier version of the server still be supported?) and [DafnyCounterExampleHandler](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyLanguageServer/Handlers/Custom/DafnyCounterExampleHandler.cs) to make them work with the new code. Add a field to [CounterExampleParams](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyLanguageServer/Handlers/Custom/CounterExampleParams.cs) that can be used to set the recursion depth when extracting fields. I (somewhat arbitrary) set this constant to `5` by default but it should be possible to add this as a setting to VS Code plugin.
6) Over 30 new tests. Comments. Stylistic changes to conform with the recently introduced style guide. 

I am happy to discuss the finer details and add any new features you think would be appropriate. Thank you!

# How to try this out?

1) Compile this Dafny fork to generate the necessary binaries.
2) Install the latest VS Code plugin for Dafny and edit the settings file to make all paths point to the respective binaries.
3) Enjoy! ...or let me know if something doesn't work - I will be grateful for any feedback!

Dargones commented:
I would be very grateful for any input from @camrein as I understand you have been working on the server and the VS Code plugin a lot!

Dargones commented:
Thank you for the feedback during the demo yesterday, @RustanLeino! I now make use of named datatype destructors and employ `:=` instead of `=>` when specifying field values. 

I am yet to support concise display of sequences for which all elements are known. I think it would be best to do so when introducing maps since both changes would require sub-classing `DafnyModelVariable`. Should I make this part of this PR or submit a follow-up?

parno commented:
It's exciting to see this revived/improved.  In another context, we brainstormed some examples that seemed like the debugger should be able to help with, but that mostly didn't work well with the old version.  I'd be curious how well they do on your new-and-improved version.  I suspect `T_int_quantifier0` and `T_int_quantifier1` still won't work, but hopefully the rest will now work?

```
method T_int_quantifier0()   
{
  assert forall i:int :: i > 0;  
}

method T_int_quantifier1(x:int)   
{
  assert forall i:int :: i > x;
}

method T_int_quantifier2()   
{
  forall i:int  
    ensures i > 0 {

    }
}

function method f(x:int) : bool
method T_int_func(x:int, y:int)   
{
  assert f(x) == f(y);  
}

method T_seq(x:seq<int>, y:seq<int>)   
{
  assert x == y;  
}

method T_array(x:array<int>, y:array<int>)   
{
  assert x == y;  
}

datatype Hand = Left(x:int, y:bool) | Right(a:int, b:int)

method T_datatype0(h0:Hand, h1:Hand)
{
  assert h0 == h1; 
}

method T_datatype1(h0:Hand, h1:Hand)
  requires h0.Left? && h1.Left?
{
  assert h0 == h1; 
}

datatype Many = C1(x:int) | C2(x:int) | C3(x:int) | C4(x:int) | C5(x:int)
function GetX(m:Many) : int { m.x }
predicate GoodMany(m:Many)
{
  match m 
    case C1(x) => x > 0
    case C2(x) => x > 0
    case C3(x) => true
    case C4(x) => x > 0
    case C5(x) => x > 0
}

method T_datatype2(m:Many)
  requires GoodMany(m)
{
  assert GetX(m) > 0;
}

method T_map0(m:map<int,int>, key:int, msg:int) 
{
    var m' := m[key := msg];    
    assert m.Values + {msg} == m'.Values;  
}

method T_map1(m:map<int,int>, key:int, msg:int) 
{
    var m' := m[key := msg];    

    if !(key in m.Keys) {
      assert m'.Values == m.Values + {msg};  
    } else {
      // Fails b/c there might be another key key' where m[key'] == m[key]
      assert m'.Values == m.Values - {m[key]} + {msg};  
    }
}
```

Dargones commented:
Thank you for this test suite, @parno ! Here is the breakdown of all these cases. Let me know what you think!

1) `T_int_quantifier0` and `T_int_quantifier1` do work. The old version did actually have code for extracting skolem constants but for some reason it was never used - I have fixed that. `T_ind_quantifier2` doesn't work though - will look into it (do you have any ideas what I should focus on to fix this?)

<img width="1540" alt="Screen Shot 2021-08-06 at 2 02 34 PM" src="https://user-images.githubusercontent.com/21084554/128570753-2897c342-7673-4a92-81d4-c5a79e9654cb.png">

2) Counterexamples for`T_int_func` and `T_seq` seem adequate to me. Not so sure about `T_array`. Arrays are compared by reference, as far as I understand, so the counterexample is technically correct, but perhaps I should make it clearer that the two variables point to different objects? Let me know.
<img width="1540" alt="Screen Shot 2021-08-06 at 2 06 14 PM" src="https://user-images.githubusercontent.com/21084554/128571721-50957ceb-46cf-4f83-a520-5a43d1cfbcce.png">

3) I think all of the counterexamples for datatypes are correct (I added `T_datatype0_1`). I don't seem to be able to extract `x` as the name of the destructor in the last case and use the index (`[0]`) instead, but this should be easy to fix. Will do that!
<img width="1540" alt="Datatypess" src="https://user-images.githubusercontent.com/21084554/128572069-545ff1f0-d587-4d3c-a8ef-6569f8bacd4b.png">
<img width="1540" alt="Diff" src="https://user-images.githubusercontent.com/21084554/128572479-aa102eaf-5274-4026-9513-220535e20686.png">

4) I don't support maps yet, but am working on this and will update this once I have counterexamples for the two last tests.

parno commented:
Wow, I'm impressed so much is working already!  That's a significant improvement.  In response to your questions above:
1. I vaguely remember Rustan saying that the forall statement might be missing a `assume {:captureState} true` inside the body.  If that's the case, perhaps fixing that would help?
2. Indeed, your understanding is correct, but I agree that it would be nice if the displayed counter-example somehow indicated the pointers were pointing at different heap locations.
3. Yep, those look quite nice!
4. Cool, looking forward to it!

camrein reviewed (commented):
Sorry for the late reply. I was on vacation and only just got back.
You are right; most of the language server-related code is by me, although Rustan created the code to retrieve the counterexamples from the more recent boogie versions.
While reviewing your changes, I've tried to concentrate on the language server-related ones, but I commented on other things that I've seen when quickly jumping through the other changes.
If you wish, I'll happily elaborate my proposals in more detail and/or dive deeper into the other changes as well.

Dargones commented:
Thank you so much @parno and @camrein for your feedback! I have addressed most of the outstanding issues in the latest batch of commits. Here is a breakdown.

1) I have adopted all the code-specific changes proposed by @camrein . Special thanks for confirming that parallel requests should not be an issue in the case of `CounterExampleLoader`!

2) It should now be clearer if two variables point to the same object (array) or not. Here are two methods that showcase the difference:
<img width="816" alt="NewArray" src="https://user-images.githubusercontent.com/21084554/128690220-2da8c4c4-cec9-4cfe-a1cd-1ffc008ff587.png">


3) Whenever all elements of a sequence are known, the sequence is now displayed with a shorthand notation. Compare how `substring` and `seq` are displayed in the counterexample to line 14:
<img width="821" alt="NewSeq" src="https://user-images.githubusercontent.com/21084554/128690301-2300a601-1aef-4ec3-8832-87c3713d8ae7.png">


4) All destructor names are now correctly extracted, including those in the examples above.

5) I have added support for maps, but have also run into some issues in the process and would be very grateful for any hint as to what to do with these. As far as I understand, these issues seem to stem from (a) the fact that Dafny cannot (in general case) compute cardinality of a set and (b) the fact that Dafny does not seem to be always able to tie together the map and the corresponding `Map#Values` and `Map#Domain` sets without help. This means that maps can and often do contain random elements in their domain. One can still get meaningful counterexamples for simple cases (see `mapUpdate` below) but I don’t think the tool works well with the two examples proposed above (I split these in three for convenience below). Or perhaps it is not supposed to? If this doesn't seem to work, I can also easily rollback map support since it is currently the last commit. The `? := ? [+2]` notation stands for "there are three distinct key-value pairs in the map for which the prover provides neither the key nor the value."
<img width="1441" alt="NewMaps" src="https://user-images.githubusercontent.com/21084554/128690521-6e5a0098-9270-4540-9b68-8ec40a50efb3.png">


6) If making `T_int_quantifier2` work requires adding `captureState` to the `DafnyPrelude`, I would probably separate this problem into a separate pull request. In the meantime, one should be able to achieve a similar effect via a helper function:
<img width="821" alt="NewHelper" src="https://user-images.githubusercontent.com/21084554/128690527-33d03d98-64d8-42f7-b5bf-a14c187d1e61.png">

Dargones commented on Source/DafnyLanguageServer/Handlers/Custom/CounterExampleParams.cs:
```diff
@@ -8,6 +8,8 @@ public class CounterExampleParams : ITextDocumentIdentifierParams, IRequest<Coun
     public TextDocumentIdentifier TextDocument { get; set; }
+    
+    public int CounterExampleDepth { get; set; } = 5;
 #pragma warning restore CS8618 // Non-nullable field must contain a non-null value when exiting constructor.
```
I have now moved the `CounterExampleDepth` after `#pragma.

Dargones commented on Source/DafnyLanguageServer/Handlers/Custom/DafnyCounterExampleHandler.cs:
```diff
@@ -29,7 +28,7 @@ public class DafnyCounterExampleHandler : ICounterExampleHandler {
         return Task.FromResult(new CounterExampleList());
       }
-      return Task.FromResult(new CounterExampleLoader(_logger, document, cancellationToken).GetCounterExamples());
+      return Task.FromResult(new CounterExampleLoader(_logger, document, cancellationToken).GetCounterExamples(request.CounterExampleDepth));
```
This makes a lot of sense! Made the corresponding change in the next to last commit.

Dargones commented on Source/DafnyLanguageServer/Handlers/Custom/DafnyCounterExampleHandler.cs:
```diff
@@ -49,52 +48,52 @@ private class CounterExampleLoader {
         }
         var counterExamples = GetLanguageSpecificModels(_document.SerializedCounterExamples)
-          .SelectMany(GetCounterExamples)
+          .SelectMany(x => GetCounterExamples(x, maxDepth))
```
I agree that given `x` a meaningful name would help - have renamed all such `x`'s here and in other places in my code.

Dargones commented on Source/DafnyLanguageServer/Handlers/Custom/DafnyCounterExampleHandler.cs:
```diff
@@ -49,52 +48,52 @@ private class CounterExampleLoader {
 
-      private LanguageModel GetLanguagSpecificModel(Model model) {
+      private DafnyModel GetLanguagSpecificModel(Model model) {
         // TODO Make view options configurable?
```
I removed the TODO

Dargones commented on Source/DafnyServer/CounterExampleProvider.cs:
```diff
@@ -25,19 +23,19 @@ public class CounterExampleProvider {
         var output = wr.ReadToEnd();
         var models = ExtractModels(output);
-        _languageSpecificModels = BuildModels(models);
+        _dafnyModels = BuildModels(models);
```
Makes sense! Made the variable local and simplified the method in general.

Dargones commented on Source/DafnyServer/CounterExampleProvider.cs:
```diff
@@ -25,19 +23,19 @@ public class CounterExampleProvider {
 
-    private List<LanguageModel> BuildModels(List<Model> modellist) {
-      var list = new List<LanguageModel>();
+    private List<DafnyModel> BuildModels(List<Model> modellist) {
```
I rewrote the method so that it now returns an `IEnumerable` - it should be possible to get generator out of this.

Dargones commented on Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs:
```diff
@@ -0,0 +1,117 @@
+        string result = state.Model.CanonicalName(Element, state);
+        if (children.Count == 0)
+          return result == "" ? "()" : result;
+        List<Tuple<string, string>> childList = new();
```
I didn't know about Named Tuples (new to C#) - made the change. Looks much better this way!

Dargones commented on Source/DafnyServer/CounterExampleProvider.cs:
```diff
@@ -58,30 +56,26 @@ public class CounterExampleProvider {
+          AddLineInformation(counterExampleState, state.FullStateName);
 
-          foreach (var variableNode in state.Vars) {
+          var vars = state.ExpandedVariableSet(2);
```
Done

Dargones commented on Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:
```diff
@@ -26,55 +28,55 @@ public class CounterExampleTest : DafnyLanguageServerTestBase {
-{
-}
-".TrimStart();
+      method Abs(x: int) returns (y: int)
```
Changed the formatting back to how it was to conform with the other tests

Dargones commented on Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:
```diff
@@ -84,25 +86,618 @@ method Abs(x: int) returns (y: int)
+    }
+
+    [TestMethod]
+    public async Task RealFull() {
```
I went through the test method names and changed them whenever they were to cryptic. This particular method should have been (and now is) named `WholeNumberAsReal` since it tests the case when a whole number is represented by a value of type `real`.

camrein commented:
Looks good to me. Feel free to drop a note if you'd like another review from me.
Just some small last input from me:

```cs
        var counterExamples = GetLanguageSpecificModels(document.SerializedCounterExamples)
          .SelectMany(model => GetCounterExamples(model))
          .ToArray();
```

In cases like `.SelectMany(model => GetCounterExamples(model))` where you pass a single input parameter through the lambda you can use a method reference/delegate instead. For example:

```cs
        var counterExamples = GetLanguageSpecificModels(document.SerializedCounterExamples)
          .SelectMany(GetCounterExamples)
          .ToArray();
```

Since there's some beefy information now available through the counterexamples, I'm wondering if there wouldn't be a better way to display this information.

Dargones commented:
Thank you, @camrein . I replaced lambdas with method groups whenever possible in my latest commit.

I think the ideal way to display counterexamples would be to take advantage of the fact that they are trees (once you start recursively expanding fields) and to use something similar to a Debug window in modern IDEs. My guess is that this would require making changes to the VSCode plugin but I also don't know how difficult this might be to implement since I never wrote a plugin myself. Do you have an idea where one would need to start \ how much effort this could entail?

RustanLeino commented:
Thanks looks really good! I just have a few comments on the output syntax, trying to make it look more like Dafny, when possible.

* I suggest removing the suffixes like `bv4` for bitvector values, since they are not part of Dafny syntax (and not printing such a suffix does not seem to lead to any ambiguity)
* Sets are printed with some elements known to be in the set and some elements known not to be in the set. I like that. There's no Dafny syntax for that, so the `:= true` and `:= false` are fine. But I'm thinking that enclosing the elements in curly braces rather than round parentheses may make it feel more set-like. For example, `{'a' := false, '!' := true}` instead of `('a' := false, '!' := true)`. What do you think?
* When all of the elements of a sequence are known then I suggest they be printed enclosed in square brackets (like in Dafny), instead of round parentheses.
* As we just discussed in person, names like `i#0@@0` can perhaps be shorted to `i` in some cases. (Historical note: the Boogie Verification Debugger suppressed variables whose name contained a `@@`. It is curious to me that there's a reason to print this variable at all. It's probably due to that Dafny has changed over the years and has not stuck to the convention that `@@`-variables are never interesting to Dafny users.)

Dargones commented:
Thank you, Rustan! Just made these changes:
<img width="697" alt="Screen Shot 2021-08-10 at 2 25 08 PM" src="https://user-images.githubusercontent.com/21084554/128937405-442fca1c-7065-4445-b56a-e08296956264.png">

Dargones commented:
I added indices to "?" (values that are not specified by the model). This helps distinguishing between situations like this:
<img width="846" alt="Screen Shot 2021-08-10 at 4 04 35 PM" src="https://user-images.githubusercontent.com/21084554/128947367-97bbe097-c582-4910-8b68-6f96c49c5118.png">
